File: String/214. Shortest Palindrome.cpp
/*
214. Shortest Palindrome


Given a string s, you are allowed to convert it to a palindrome by adding characters in front of it. Find and return the shortest palindrome you can find by performing this transformation.

Example 1:

Input: "aacecaaa"
Output: "aaacecaaa"
Example 2:

Input: "abcd"
Output: "dcbabcd"

dcbaabcd



*/

/*
解法思路： 
把这个题目转化下： 其实是在问从index 0开始，部分最长的palindrome, 
如果我们得到这个最长的palidrome了, 可以reverse这个s内部分最长的palindrome 后面的，构成全体最短的palidrome 

比如 catacbd, 单词内部分最长的palidrome 是catac, 所以把部分最长的palidrom后面部分reverse 就构成了我们想要的最短的 dbcatacbd

可以利用kmp找到以index 0 开始最长的palindrome，
用 s + "#" + reverse(s) 这样从0开始最长的palidrome 就会在最后一个prefix显示出来，因为如果是最长的，会以#对称 + 头尾一样相等 （prefix即是suffix)，
比如catac, reverse 还是catac, 所以会以# 对称，并且相等，但是abc, reverse cba, 

最后一位肯定是部分最长的suffix（prefix即是suffix， 头尾相等，且掉个也相等）

input:

catacb

Temp String:

catacb # bcatac

KMP table:

c a t a c b # b c a t a c

0 0 0 0 1 0 0 0 1 2 3 4 5
*/


class Solution {
public:
    string shortestPalindrome(string s) {
        string rev = string(s.rbegin(), s.rend());
        vector<int>prefix = getPrefix(s+"#"+rev);
        string end = s.substr(prefix[prefix.size()-1]);
        reverse(end.begin(),end.end());
        return end+s;
    }

    vector<int>getPrefix(string && s){
        vector<int>vec(s.size());
        int j = 0;
        for(int i = 1; i<s.size();i++){
            while(j>0 && s[j]!=s[i])
                j = vec[j-1];
            if(s[i] == s[j])
                j++;
            vec[i] = j;
        }
        return vec;
    }
};


/*
不用#
先用kmp 求整个s的prefix array 

然后两个pointer， 一个pointer从s最后出发， 一个记录match的个数，如果s[i] == s[match] match++
如果match > 0 && s[i] != s[match] 不动i，动match回到上一个即是prefix 又是suffix的
比如 a b a b c... a b a b a
            -    -
一个在c，一个在a不match， 把match = prefix[match - 1] = 2, 发现match继续寻找
比如 a b a b c... a b a b a
        -        -
*/

class Solution {
public:
    string shortestPalindrome(string s) {
        vector<int>prefix = getPrefix(s);
        int maxPalLen = getmaxPalindrome(prefix, s);
        string nonPalindrome = s.substr(maxPalLen);
        reverse(nonPalindrome.begin(), nonPalindrome.end());
        return nonPalindrome+s;
    }

    vector<int>getPrefix(string & s){
        vector<int>vec(s.size());
        int j = 0;
        for(int i = 1; i<s.size();i++){
            while(j>0 && s[j]!=s[i])
                j = vec[j-1];
            if(s[i] == s[j])
                j++;
            vec[i] = j;
        }
        return vec;
    }

    int getmaxPalindrome(const vector<int>&prefix, const string &s){
        int match = 0;
        for(int i = s.size()-1; i>=0; i--){
            while(match > 0 && s[match]!= s[i]) match = prefix[match-1];
            if(s[match] == s[i]) match++;
        }
        return match;
    }
};



class Solution {
public:
    string shortestPalindrome(string s) {
        string snew = "$#";
        for(int i = 0; i<s.size();i++){
            snew += s[i];
            snew += "#";
        }
        snew += "^";
        int mxlen = manacher(snew);
        string nonPal = s.substr(mxlen);
        reverse(nonPal.begin(), nonPal.end());
        return nonPal + s;
    }

    int manacher(const string &snew){
        int id = 0, mx = 0;
        int mxlen = 1;
        vector<int>p(snew.size(), 0);
        for(int i = 1; i<snew.size()-1; i++){
            if(i<mx){
                p[i] = min(p[id*2 - i], mx - i);
            }else 
                p[i] = 1;
            while(snew[i-p[i]] == snew[i+p[i]])
                p[i]++;
            if(p[i] + i > mx){
                mx = p[i] + i;
                id = i;
                if(i - p[i] == 0) mxlen = p[i];
            }
        }
        return mxlen-1;
    }
};

File: kmp/214. Shortest Palindrome.py
"""
214. Shortest Palindrome

Given a string s, you are allowed to convert it to a palindrome by adding characters in front of it.
Find and return the shortest palindrome you can find by performing this transformation.

Example 1:

Input: "aacecaaa"
Output: "aaacecaaa"
Example 2:

Input: "abcd"
Output: "dcbabcd"
"""

class Solution:
    def shortestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        snew = s + "#" +s[::-1]
        def getPrefix(snew):
            prefix = [0]*len(snew)
            j = 0
            for i in range(1,len(snew)):
                while j > 0 and snew[i]!=snew[j]:
                    j = prefix[j-1]
                if snew[i] == snew[j]: j+=1
                prefix[i] = j
            return prefix
        kmp = getPrefix(snew)
        nonpal = s[kmp[-1]:]
        return nonpal[::-1] + s


import functools
class Solution:
    def shortestPalindrome(self, s):
        snew = functools.reduce(lambda x, y: x + y + "#", s,"$#")+"^"
        def manacher(snew):
            p = [0]*len(snew)
            mx = id = maxlen = 0
            for i in range(1,len(snew)-1):
                if i < mx: 
                    p[i] = min(p[id*2 - i], p[mx - i])
                else: 
                    p[i] = 1
                while snew[i + p[i]] == snew[i - p[i]]:
                    p[i] += 1
                if p[i] + i > mx:
                    id, mx = i, p[i]+i
                if p[i] == i:
                    maxlen = max(maxlen, p[i]-1)
            return maxlen
        maxlen = manacher(snew)
        nonpal = s[maxlen:]
        return nonpal[::-1] + s

"""

The basic idea is to find the longest palindrome starting from s[0], so that fewest charactors are needed in front of s.

For any charactor(c) in s appearing more than once, define l and r as the first and last index of c, 
then the max length of palindrome starting from s[0] would be no larger than l+r+1, or the first c could never be matched.

For any charactor(c) in s appearing just once, l=r. Here is another key optimization.if s[:l+r+1] is not a palindrome, 
the max length of palindrome starting from s[0] would be no larger than l. 
In other words, c must be excluded from the palindrome, or it could not be matched. 
If s[:l+r+1] is a palindrome, actually c is the center, matched by itself.

"""


class Solution:
    def shortestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        if not s: return ''
        k=len(s)
        for c in set(s):
            l,r=s.find(c),s.rfind(c)
            k=min(k,l if l==r and s[:l+r+1]!=s[:l+r+1][::-1] else l+r+1)
        for i in range(k,0,-1):
            if s[:i]==s[:i][::-1]: return s[i:][::-1]+s
